### PR TITLE
Dmx input esp dmx

### PR DESCRIPTION
**Pull request by netmindz**
Add support for listening to DMX, just like e1.31 and art-net but over a wired connection.

Note: needs the newer ESP32 platform so only enabled for the _V4 builds

### CODE REVIEW

**Comment by netmindz** [comment]
Please test using the esp32_4MB_V4_S build target

**Comment by ewoudwijma** [comment]
I see a compile error on:

image

Could you compile this separately and see what happens? Most likely over 100% flash size...

@softhack007 , should we change esp32_4MB_PSRAM_M to esp32_4MB_PSRAM_S (small) ? Or another way PSRAM is partitioned possible?

**Comment by softhack007** [comment]
> Could you compile this separately and see what happens? Most likely over 100% flash size...
> 
> @softhack007 , should we change esp32_4MB_PSRAM_M to esp32_4MB_PSRAM_S (small) ? Or another way PSRAM is partitioned possible?

Yes, I think that's necessary, as we are regularly fighting with size on the PSRAM config. Another option (but incompatible with previous installs) would be to use another partition file for that config - it would only have 256Kb for file system (LittleFS), but adds about 300Kb of program space.

I'll look into the compile errors 👍

**Comment by netmindz** [comment]
Yeah could squeeze a little more space for the app Vs file system that would be great. Don't think we really need much filesystem space for most users do we?

**Comment by softhack007** [comment]
You can already do that for testing. I also gives you more space for debug build. Just uncomment line 263 in platformio.ini, and comment out line 262.

https://github.com/MoonModules/WLED/blob/922b8b69dbc0145b70be6c61cb73968dde254d0d/platformio.ini#L262-L263

Downside is, it leads to reformatting the flash, so you'll lose your config when uploading firmware with a different partitioning.

I'm just changing the PSRAM buildenv to "_S" - lets see if that's enough.

**Comment by softhack007** [comment]
@netmindz technical question - are you relying on the SparkFun DMX driver (in wled dependancies folder), or using a different driver?

I'm hoping for the second, because the SparkFun driver does a few suspicious things (like timer interrupts on Serial2), and its not compatible with ESP32-S2 or ESP32-C3.

**Comment by softhack007** [comment]
> @softhack007 , should we change esp32_4MB_PSRAM_M to esp32_4MB_PSRAM_S (small) ? Or another way PSRAM is partitioned possible?

Just did both:
* [env:esp32_4MB_PSRAM_S] downgraded from "M" to "S" (93% flash used)
* [env:esp32_4MB_PSRAM_M] added, with smaller partition for file system (84% flash used)

see my latest commit e8d3712d4f4e645e7cde3f476e138a3044371efe

**Comment by netmindz** [comment]
I'm using [esp_dmx](https://github.com/someweisguy/esp_dmx) - which does use also uses interrupts and Serial2, I don't have any of the -S2, -S3 or -C3 board to test with sadly.

**Comment by netmindz** [comment]
Thanks, @softhack007 that's the builds all running green now.

**Comment by ewoudwijma** [comment]
Is this ready to merge? So can be part of the release ...

**Comment by netmindz** [comment]
I've tested and seems to work for me, was trying to get some other folk to test as well just to be sure.

As it's only enabled for the experimental _V4 builds it should be relatively low risk

File: platformio.ini [hunk 1]
```diff
@@ -289,6 +289,7 @@ lib_depsV4 =
   ;https://github.com/Makuna/NeoPixelBus.git#master ;; NPB 2.6.9 tends to crash whith IDF V4.4.3 -> use latest NeoPixelBus dev instead
   makuna/NeoPixelBus @ 2.7.1
   https://github.com/pbolduc/AsyncTCP.git @ 1.2.0
+  https://github.com/someweisguy/esp_dmx.git#feature/rdm-rx
```
Comment (MoonModules) [inline platformio.ini:293]: Are you sure we need to enable this for all v4 bins? I can imagine this needs to be only added for bins where we want this feature enabled. @softhack007 what do you think?

Comment (netmindz) [reply platformio.ini:293]: Unless you define the pin, then the feature is disabled. Depends on how we want to play the balance between having features available vs needing to dig about in the build flags in order to keep the image size down. It's not really a board-specific feature, so don't really want to duplicate all the targets with a _with_dmx_input value.
DMX over IP reception is in all builds by default already, this is just over serial. This is why I haven't made it as a usermod as it's just an extension of the core DMX feature

Comment (MoonModules) [reply platformio.ini:293]: Okay, was just wondering if https://github.com/someweisguy/esp_dmx.git#feature/rdm-rx is compiled in, also when the pin is not defined, in that case taking up flash space, but maybe that isn't the case.

Will merge it into mdev, if library takes up space we can always optimise that later

**Pull request merged**